### PR TITLE
[WIP] Add Tpetra tests.

### PR DESCRIPTION
**Pull request by bangerth**
This is definitely a WIP patch: 284 tests for the Tpetra wrappers, or which currently 41 succeed. I created these tests by cloning the existing `tests/trilinos/` directory and replacing `TrilinosWrappers::` by `LinearAlgebra::TpetraWrappers::` plus some minor adjustments such as adding template arguments. I'm sure many tests need additional adjustments.

Many (=the vast majority) of these tests clearly do not currently work. But the patch may help us isolate what needs to be fixed to make them work (such as the issue with `VectorReference::operator Number()` mentioned in #16610) and I will rebase this branch frequently to see how things proceed.

I'll also keep a list of things that need to be fixed for a substantial number of tests, and that I'll keep updated as I see things. Here's a start:

- [x] We need an implementation of `VectorReference::operator Number()`
- [x] `SparseMatrix::operator() (int,int)` does not exist
- [x] `SparseMatrix::el(int,int)` does not exist
- [x] `Vector::operator==` is missing.

@kinnewig @jpthiele FYI

### CODE REVIEW

**Comment by jpthiele** [comment]
Very nice! 

A few questions for the direct solver tests: 
- will we be able to add a full setup to the test server, so we can test whether all the solvers supported by Amesos2 work? 
That would possibly be its own test that is only run there as it would issue various Exceptions when trying 
to create a solver that is not configured in Trilinos
- should we also have an exception test to see that all the warnings are correctly raised, so only KLU2 works as
that is the only one bundled with Amesos2
- Alternative option: is there a nice way to decide on the file the output should be compared with based on the CMake configuration? Possibly with a `direct_solver.output.in` that is configured by CMake based on what it can find?
-  We should still try to add our subset of tests in our PRs according to this scheme here, right?

**Comment by bangerth** [comment]
Yes, with have a number of tests that have outputs like `tests/trilinos/precondition_muelu_smoother.with_trilinos_with_muelu=on.with_64bit_indices=off.output` that express requirements. Whether it's worth testing that we get errors on unsupported configurations is perhaps debatable, but at least we can limit running tests if they require a configuration not currently available.

**Comment by bangerth** [comment]
With #16613, we are now at 55/284 tests succeeding.

**Comment by bangerth** [comment]
And with #16616, we're now at 71/284 tests succeeding. There's progress :-)

**Comment by jpthiele** [comment]
> Yes, with have a number of tests that have outputs like `tests/trilinos/precondition_muelu_smoother.with_trilinos_with_muelu=on.with_64bit_indices=off.output` that express requirements. Whether it's worth testing that we get errors on unsupported configurations is perhaps debatable, but at least we can limit running tests if they require a configuration not currently available.

Okay good to know!
Then the question would be how the current setups on the tester look like 
and whether we want to have a 'feature complete' Amesos2 test setup.

**Comment by jpthiele** [comment]
With the `swap` function implemented in #16602 
at least the SolverCG test should pass since I was able to compute the reference solution for direct_solver_2.cc with it.
If you want I can make a dedicated PR to bring the `deal_solver_*.cc` test to passing.
Or two, one for swap alone and one for the tests.

**Comment by jpthiele** [comment]
We now have #16625

**Comment by bangerth** [comment]
On 2/10/24 00:07, Jan Philipp Thiele wrote:
> 
> Then the question would be how the current setups on the tester look like
> and whether we want to have a 'feature complete' Amesos2 test setup.

I don't actually know. The various testers may also have different setups. 
@tamiko -- do you know?

**Comment by kinnewig** [comment]
I like to keep using this PR to keep track of the work we have done so far.


> - [ ] We need an implementation of VectorReference::operator Number()

Implemented by: #16613


> - [ ] SparseMatrix::operator() (int,int) does not exist
> - [ ] SparseMatrix::el(int,int) does not exist

Implemented by: #16616


> - [ ] Vector::operator== is missing.

Implemented by: #16664

The current master passes 93 tests out of 284.

**Comment by bangerth** [comment]
Now at 114 out of 288 tests succeeding :-)

**Comment by bangerth** [comment]
I have to admit that I don't understand the deadlock (nor perhaps what the test does). I *think* what the test does is create a Tpetra vector locally on a process and copy it to a deal.II vector:
```
if (n_mpi_processes > 1)
    MPI_Comm_split(mpi_communicator, colour, key, &serial_communicator);
  else
    serial_communicator = mpi_communicator;

  if (this_mpi_process == 0)
    {
      const LinearAlgebra::TpetraWrappers::Vector<double> tril_vec(
        complete_index_set(10), serial_communicator);

      // Check copy constructor
      const Vector<double> local_vector_1(tril_vec);

      // Check equality operator
      Vector<double> local_vector_2;
      local_vector_2 = tril_vec;
    }
```
I believe that the `serial_communicator` we get is really just `MPI_COMM_SELF`. So we create a non-parallel vector, and then we copy it. What could actually go wrong here? We never use any MPI communicators with more than one process. I have to admit that I don't even know what this is supposed to test for the Epetra wrappers, nor do I understand how that could deadlock with the new Tpetra wrappers. Can you explain what I'm (clearly) missing?

**Comment by bangerth** [comment]
We're now at 141/288 succeeding tests.

**Comment by jpthiele** [comment]
@bangerth I'm wondering if there is another chunk of tests sizable enough to add as a separate PR again, e.g. for the BlockVector and Matrix. If I can help with that let me know!

**Comment by jpthiele** [comment]
On another note I had a quick look at the NOX tests, 
but I am not (yet) familiar enough with NOX to port them. 
Maybe someone else is?

**Comment by bangerth** [comment]
I'm not sure how much time I will have to work on this, but I did push a rebased version of this branch. Right now, on `master`, I have 10/113 tests failing (=103 succeed):
```
The following tests FAILED:
        14224 - trilinos_tpetra/deal_solver_01.debug (Failed)
        14226 - trilinos_tpetra/deal_solver_02.debug (Failed)
        14228 - trilinos_tpetra/deal_solver_03.debug (Failed)
        14230 - trilinos_tpetra/deal_solver_04.debug (Failed)
        14232 - trilinos_tpetra/deal_solver_05.debug (Failed)
        14234 - trilinos_tpetra/deal_solver_06.debug (Failed)
        14236 - test_dependency/trilinos_tpetra.direct_solver_01.debug.executable (Failed)
        14237 - trilinos_tpetra/direct_solver_01.mpirun=1.debug (Not Run)
        14240 - test_dependency/trilinos_tpetra.direct_solver_02.debug.executable (Failed)
        14241 - trilinos_tpetra/direct_solver_02.mpirun=1.debug (Not Run)
```
They seem to be failing because of this error:
```
14224: /home/fac/g/bangerth/p/deal.II/1/dealii/tests/trilinos_tpetra/deal_solver_01.cc:74: error: undefined reference to 'dealii::GrowingVectorMemory<dealii::LinearAlgebra::TpetraWrappers::Vector<double, dealii::MemorySpace::Host> >::GrowingVectorMemory(unsigned long, bool)'
```
This should be relatively easy to fix.

On the branch of this PR, I have 159/295 tests failing (=136 succeed). So there are 33 on the branch that actually succeed, but I have to leave and so can't look up right now which ones these are. The ones that fail are as follows:
```
The following tests FAILED:
        14128 - trilinos_tpetra/03b.debug (Failed)
        14134 - trilinos_tpetra/06.debug (Failed)
        14136 - trilinos_tpetra/07.debug (Failed)
        14172 - trilinos_tpetra/26.debug (Failed)
        14186 - trilinos_tpetra/33.debug (Failed)
        14196 - trilinos_tpetra/38.debug (Failed)
        14202 - trilinos_tpetra/41.debug (Failed)
        14212 - trilinos_tpetra/49a.debug (Failed)
        14214 - trilinos_tpetra/49b.debug (Failed)
        14216 - trilinos_tpetra/50.debug (Failed)
        14232 - trilinos_tpetra/58.debug (Failed)
        14234 - trilinos_tpetra/60.debug (Failed)
        14236 - trilinos_tpetra/61.debug (Failed)
        14255 - trilinos_tpetra/add_matrices_01.mpirun=3.debug (Failed)
        14259 - trilinos_tpetra/add_matrices_02.mpirun=3.debug (Failed)
        14263 - trilinos_tpetra/add_matrices_03.mpirun=3.debug (Failed)
        14266 - trilinos_tpetra/add_matrices_04.debug (Failed)
        14268 - trilinos_tpetra/add_matrices_05.debug (Failed)
        14270 - trilinos_tpetra/add_matrices_06.debug (Failed)
        14272 - test_dependency/trilinos_tpetra.amg_nondeterministic_02.debug.executable (Failed)
        14273 - trilinos_tpetra/amg_nondeterministic_02.mpirun=2.debug (Not Run)
        14276 - trilinos_tpetra/assemble_matrix_parallel_01.debug (Failed)
        14278 - test_dependency/trilinos_tpetra.assemble_matrix_parallel_02.debug.executable (Failed)
        14279 - trilinos_tpetra/assemble_matrix_parallel_02.mpirun=1.debug (Not Run)
        14282 - trilinos_tpetra/assemble_matrix_parallel_02.mpirun=4.debug (Not Run)
        14284 - test_dependency/trilinos_tpetra.assemble_matrix_parallel_03.debug.executable (Failed)
        14285 - trilinos_tpetra/assemble_matrix_parallel_03.mpirun=4.debug (Not Run)
        14288 - test_dependency/trilinos_tpetra.assemble_matrix_parallel_04.debug.executable (Failed)
        14289 - trilinos_tpetra/assemble_matrix_parallel_04.mpirun=4.debug (Not Run)
        14292 - test_dependency/trilinos_tpetra.assemble_matrix_parallel_05.debug.executable (Failed)
        14293 - trilinos_tpetra/assemble_matrix_parallel_05.mpirun=1.debug (Not Run)
        14296 - trilinos_tpetra/assemble_matrix_parallel_05.mpirun=4.debug (Not Run)
        14298 - test_dependency/trilinos_tpetra.assemble_matrix_parallel_06.debug.executable (Failed)
        14299 - trilinos_tpetra/assemble_matrix_parallel_06.mpirun=4.debug (Not Run)
        14302 - test_dependency/trilinos_tpetra.assemble_matrix_parallel_07.debug.executable (Failed)
        14303 - trilinos_tpetra/assemble_matrix_parallel_07.mpirun=4.debug (Not Run)
        14306 - test_dependency/trilinos_tpetra.block_matrices_01.debug.executable (Failed)
        14307 - trilinos_tpetra/block_matrices_01.mpirun=1.debug (Not Run)
        14310 - test_dependency/trilinos_tpetra.block_relax_no_rows.debug.executable (Failed)
        14311 - trilinos_tpetra/block_relax_no_rows.mpirun=3.debug (Not Run)
        14314 - trilinos_tpetra/block_sparse_matrix_add_01.debug (Failed)
        14316 - trilinos_tpetra/block_sparse_matrix_set_01.debug (Failed)
        14318 - trilinos_tpetra/block_sparse_matrix_vector_01.debug (Failed)
        14320 - trilinos_tpetra/block_sparse_matrix_vector_02.debug (Failed)
        14322 - trilinos_tpetra/block_sparse_matrix_vector_03.debug (Failed)
        14324 - trilinos_tpetra/block_sparse_matrix_vector_04.debug (Failed)
        14326 - trilinos_tpetra/block_vector_iterator_01.debug (Failed)
        14328 - trilinos_tpetra/block_vector_iterator_02.debug (Failed)
        14330 - trilinos_tpetra/block_vector_iterator_03.debug (Failed)
        14332 - trilinos_tpetra/block_vector_move_operations.debug (Failed)
        14334 - test_dependency/trilinos_tpetra.copy_to_dealvec_block.debug.executable (Failed)
        14335 - trilinos_tpetra/copy_to_dealvec_block.mpirun=4.debug (Not Run)
        14338 - trilinos_tpetra/deal_solver_01.debug (Failed)
        14340 - trilinos_tpetra/deal_solver_02.debug (Failed)
        14342 - trilinos_tpetra/deal_solver_03.debug (Failed)
        14344 - trilinos_tpetra/deal_solver_04.debug (Failed)
        14346 - trilinos_tpetra/deal_solver_05.debug (Failed)
        14348 - trilinos_tpetra/deal_solver_06.debug (Failed)
        14350 - trilinos_tpetra/direct_solver.debug (Failed)
        14352 - test_dependency/trilinos_tpetra.direct_solver_01.debug.executable (Failed)
        14353 - trilinos_tpetra/direct_solver_01.mpirun=1.debug (Not Run)
        14356 - test_dependency/trilinos_tpetra.direct_solver_02.debug.executable (Failed)
        14357 - trilinos_tpetra/direct_solver_02.mpirun=1.debug (Not Run)
        14360 - test_dependency/trilinos_tpetra.direct_solver_2.debug.executable (Failed)
        14361 - trilinos_tpetra/direct_solver_2.mpirun=1.debug (Not Run)
        14364 - test_dependency/trilinos_tpetra.direct_solver_3.debug.executable (Failed)
        14365 - trilinos_tpetra/direct_solver_3.mpirun=2.debug (Not Run)
        14368 - test_dependency/trilinos_tpetra.elide_zeros.debug.executable (Failed)
        14369 - trilinos_tpetra/elide_zeros.mpirun=2.debug (Not Run)
        14372 - test_dependency/trilinos_tpetra.epetra_vector_01.debug.executable (Failed)
        14373 - trilinos_tpetra/epetra_vector_01.mpirun=2.debug (Not Run)
        14376 - test_dependency/trilinos_tpetra.epetra_vector_02.debug.executable (Failed)
        14377 - trilinos_tpetra/epetra_vector_02.mpirun=2.debug (Not Run)
        14380 - test_dependency/trilinos_tpetra.epetra_vector_03.debug.executable (Failed)
        14381 - trilinos_tpetra/epetra_vector_03.mpirun=2.debug (Not Run)
        14388 - test_dependency/trilinos_tpetra.index_set_02.debug.executable (Failed)
        14389 - trilinos_tpetra/index_set_02.mpirun=2.debug (Not Run)
        14392 - test_dependency/trilinos_tpetra.mg_transfer_prebuilt_01.debug.executable (Failed)
        14393 - trilinos_tpetra/mg_transfer_prebuilt_01.mpirun=1.debug (Not Run)
        14401 - trilinos_tpetra/n_nonzeros_02_parallel.mpirun=9.debug (Failed)
        14404 - trilinos_tpetra/nox_solver_01.debug (Failed)
        14406 - trilinos_tpetra/nox_solver_02.debug (Failed)
        14408 - trilinos_tpetra/nox_solver_03.debug (Failed)
        14410 - trilinos_tpetra/nox_solver_04.debug (Failed)
        14412 - trilinos_tpetra/nox_solver_05.debug (Failed)
        14414 - trilinos_tpetra/nox_solver_06.debug (Failed)
        14416 - test_dependency/trilinos_tpetra.parallel_block_vector_copy_01.debug.executable (Failed)
        14417 - trilinos_tpetra/parallel_block_vector_copy_01.mpirun=2.debug (Not Run)
        14420 - test_dependency/trilinos_tpetra.parallel_block_vector_reinit_01.debug.executable (Failed)
        14421 - trilinos_tpetra/parallel_block_vector_reinit_01.mpirun=10.debug (Not Run)
        14426 - trilinos_tpetra/precondition.debug (Failed)
        14428 - test_dependency/trilinos_tpetra.print_01.debug.executable (Failed)
        14429 - trilinos_tpetra/print_01.mpirun=2.debug (Not Run)
        14432 - test_dependency/trilinos_tpetra.readwritevector.debug.executable (Failed)
        14433 - trilinos_tpetra/readwritevector.mpirun=2.debug (Not Run)
        14436 - test_dependency/trilinos_tpetra.readwritevector_02.debug.executable (Failed)
        14437 - trilinos_tpetra/readwritevector_02.mpirun=2.debug (Not Run)
        14440 - test_dependency/trilinos_tpetra.readwritevector_03.debug.executable (Failed)
        14441 - trilinos_tpetra/readwritevector_03.mpirun=2.debug (Not Run)
        14444 - test_dependency/trilinos_tpetra.readwritevector_04.debug.executable (Failed)
        14445 - trilinos_tpetra/readwritevector_04.mpirun=2.debug (Not Run)
        14452 - trilinos_tpetra/renumbering_01.mpirun=2.debug (Failed)
        14462 - trilinos_tpetra/solver_03.debug (Failed)
        14464 - trilinos_tpetra/solver_05.debug (Failed)
        14466 - trilinos_tpetra/solver_07.debug (Failed)
        14468 - trilinos_tpetra/solver_control_01.debug (Failed)
        14470 - trilinos_tpetra/solver_control_02.debug (Failed)
        14472 - trilinos_tpetra/solver_control_03.debug (Failed)
        14474 - trilinos_tpetra/solver_control_04.debug (Failed)
        14476 - trilinos_tpetra/solver_control_05.debug (Failed)
        14478 - test_dependency/trilinos_tpetra.solver_control_06.debug.executable (Failed)
        14479 - trilinos_tpetra/solver_control_06.mpirun=1.debug (Not Run)
        14482 - trilinos_tpetra/solver_control_06.mpirun=2.debug (Not Run)
        14484 - trilinos_tpetra/sparse_matrix_01.debug (Failed)
        14492 - trilinos_tpetra/sparse_matrix_05.debug (Failed)
        14494 - trilinos_tpetra/sparse_matrix_07.debug (Failed)
        14496 - trilinos_tpetra/sparse_matrix_07_rectangle.debug (Failed)
        14498 - trilinos_tpetra/sparse_matrix_08.debug (Failed)
        14500 - trilinos_tpetra/sparse_matrix_add_01.debug (Failed)
        14502 - trilinos_tpetra/sparse_matrix_add_02.debug (Failed)
        14505 - trilinos_tpetra/sparse_matrix_add_03.mpirun=3.debug (Failed)
        14508 - trilinos_tpetra/sparse_matrix_copy_from_01.debug (Failed)
        14514 - trilinos_tpetra/sparse_matrix_iterator.debug (Failed)
        14518 - test_dependency/trilinos_tpetra.sparse_matrix_iterator_02.debug.executable (Failed)
        14519 - trilinos_tpetra/sparse_matrix_iterator_02.mpirun=1.debug (Not Run)
        14522 - trilinos_tpetra/sparse_matrix_iterator_02.mpirun=2.debug (Not Run)
        14529 - trilinos_tpetra/sparse_matrix_set_03.mpirun=1.debug (Failed)
        14532 - trilinos_tpetra/sparse_matrix_set_03.mpirun=2.debug (Failed)
        14550 - trilinos_tpetra/sparse_matrix_vector_09.debug (Failed)
        14552 - trilinos_tpetra/sparse_matrix_vector_10.debug (Failed)
        14554 - trilinos_tpetra/sparse_matrix_vector_11.debug (Failed)
        14556 - trilinos_tpetra/sparse_matrix_vector_12.debug (Failed)
        14558 - trilinos_tpetra/sparse_matrix_vector_13.debug (Failed)
        14572 - trilinos_tpetra/sparsity_pattern_04.debug (Failed)
        14574 - trilinos_tpetra/sparsity_pattern_05.debug (Failed)
        14577 - trilinos_tpetra/sparsity_pattern_06.mpirun=3.debug (Failed)
        14580 - test_dependency/trilinos_tpetra.sparsity_pattern_07.debug.executable (Failed)
        14581 - trilinos_tpetra/sparsity_pattern_07.mpirun=1.debug (Not Run)
        14584 - trilinos_tpetra/sparsity_pattern_07.mpirun=2.debug (Not Run)
        14586 - trilinos_tpetra/step-77-with-nox.debug (Failed)
        14588 - test_dependency/trilinos_tpetra.subtract_mean_value_04.debug.executable (Failed)
        14589 - trilinos_tpetra/subtract_mean_value_04.mpirun=2.debug (Not Run)
        14606 - trilinos_tpetra/trilinos_64_bit_crash_01.debug (Failed)
        14632 - trilinos_tpetra/trilinos_sparsity_pattern_01.debug (Failed)
        14634 - test_dependency/trilinos_tpetra.trilinos_sparsity_pattern_02.debug.executable (Failed)
        14635 - trilinos_tpetra/trilinos_sparsity_pattern_02.mpirun=4.debug (Not Run)
        14639 - trilinos_tpetra/trilinos_sparsity_pattern_03.mpirun=4.debug (Failed)
        14643 - trilinos_tpetra/trilinos_sparsity_pattern_04.mpirun=3.debug (Failed)
        14647 - trilinos_tpetra/trilinos_sparsity_pattern_05.mpirun=3.debug (Failed)
        14650 - trilinos_tpetra/trilinos_sparsity_pattern_iterators_01.debug (Failed)
        14652 - trilinos_tpetra/trilinos_sparsity_pattern_iterators_02.debug (Failed)
        14655 - trilinos_tpetra/trilinos_sparsity_pattern_iterators_03.mpirun=1.debug (Failed)
        14658 - trilinos_tpetra/trilinos_sparsity_pattern_iterators_03.mpirun=2.debug (Failed)
        14660 - trilinos_tpetra/update_ghosts.debug (Failed)
        14678 - trilinos_tpetra/vector_local_copy_01.mpirun=2.debug (Failed)
        14680 - trilinos_tpetra/vector_move_operations.debug (Failed)
        14688 - test_dependency/trilinos_tpetra.vector_reinit_to_linear_map.debug.executable (Failed)
        14689 - trilinos_tpetra/vector_reinit_to_linear_map.mpirun=2.debug (Not Run)
        14695 - trilinos_tpetra/vector_swap_02.mpirun=2.debug (Failed)
```
In other words: We should fix the ten failing on master, and merge the 33 that are succeeding on the branch.

**Comment by jpthiele** [comment]
On it!

As a side note, the tests `directsolver(2|3)?` can be deleted as they got a leading 0 and a 01

**Comment by bangerth** [comment]
On 7/18/24 11:29, Jan Philipp Thiele wrote:
> As a side note, the tests |directsolver(2|3)?| can be deleted as they got a 
> leading 0 and a 01

Feel free to do this in a separate PR.

**Comment by jpthiele** [comment]
> On 7/18/24 11:29, Jan Philipp Thiele wrote: As a side note, the tests |directsolver(2|3)?| can be deleted as they got a leading 0 and a 01
> Feel free to do this in a separate PR.

Maybe I should have added that currently the ones with old names are only in this PR branch, 
so nothing to do on master.

**Comment by jpthiele** [comment]
Interestingly, the deal_solver and direct_solver_0X tests passed both on my laptop and in the CI run of #17292,
so nothing to do there maybe. Or at least nothing I could search for. 

Also, dangerously, the amg and muelu tests are not part of the failing set as they are still set to 32Bit indices only. 
Or Trilinos with MueLu, which is also only set to on with 32Bit indices.
